### PR TITLE
gh-111972: unicode: Make `ucnhash_capi`capsule initialization thread-safe

### LINKED ISSUE #111972: unicode: make ucnhash_capi initialization thread-safe in `--disable-gil` builds
# Feature or enhancement

The `_PyUnicode_Name_CAPI` provides functions to get the name for a given Unicode character code and vice versa. It is lazily initialized and stored in the per-interpreter `_Py_unicode_state`:

https://github.com/python/cpython/blob/3932b0f7b1566374427daa8bc47203032015e350/Include/internal/pycore_unicodeobject.h#L425-L432

The initialization of the `ucnhash_capi` isn't thread-safe without the GIL. (There can be a data race on reading and writing `ucnhash_capi`).

Mostly for my own reference, here are the similar modifications in the `nogil-3.12` fork: https://github.com/colesbury/nogil-3.12/commit/5d006db9fa

### Linked PRs
* gh-112249

### PR DESCRIPTION
**Pull request by Eclips4**
* Issue: gh-111972

### CODE REVIEW

File: Python/codecs.c [hunk 1]
```diff
@@ -954,13 +952,13 @@ PyObject *PyCodec_NameReplaceErrors(PyObject *exc)
...
-        if (!ucnhash_capi) {
-            /* load the unicode data module */
-            ucnhash_capi = (_PyUnicode_Name_CAPI *)PyCapsule_Import(
-                                            PyUnicodeData_CAPSULE_NAME, 1);
-            if (!ucnhash_capi) {
-                return NULL;
-            }
+        _PyUnicode_Name_CAPI *ucnhash_capi = _PyUnicode_GetNameCAPI();
+        if (ucnhash_capi == NULL) {
+            PyErr_SetString(
```
Comment (colesbury) [inline Python/codecs.c]: I don't think this error message is appropriate here. We should probably just keep the existing behavior (just `return NULL`) and propagate up the `PyCapsule_Import` error.

**Review by colesbury — COMMENTED** [review]
I think we also need to handle the `case 'N'` in `_PyUnicode_DecodeUnicodeEscapeInternal`.

File: Python/codecs.c (hunk 1 above)
Comment (Eclips4) [reply Python/codecs.c]: Done.

**Comment by Eclips4** [comment]
Should we do something with `_PyUnicode_Fini`?
https://github.com/python/cpython/blob/14fd86a59d0d91fe72641efeb14a59d99127dec3/Objects/unicodeobject.c#L15445-L15460

File: Lib/test/test_unicodedata.py [hunk 2]
```diff
@@ -289,8 +289,7 @@ def test_failed_import_during_compiling(self):
         # We use a separate process because the unicodedata module may already
         # have been loaded in this process.
         result = script_helper.assert_python_failure("-c", code)
-        error = "SyntaxError: (unicode error) \\N escapes not supported " \
-            "(can't load unicodedata module)"
+        error = 'ImportError: PyCapsule_Import could not import module "unicodedata"'
```
Comment (colesbury) [inline Lib/test/test_unicodedata.py]: I think we want to keep the previous error message. The message about `\N` escapes makes sense in `_PyUnicode_DecodeUnicodeEscapeInternal` in `Objects/unicodeobject`.

**Review by colesbury — COMMENTED** [review]
`_PyUnicode_Fini` looks fine. I don't think it needs any special handling.

File: Lib/test/test_unicodedata.py (hunk 2 above)
Comment (Eclips4) [reply Lib/test/test_unicodedata.py]: Sounds reasonable.
Done.

**Review by colesbury — APPROVED** [review]

**Comment by colesbury** [comment]
@erlend-aasland, would you please review this when you get a chance?

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks for the PR, Kirill, and thanks for the review, Sam!

File: Objects/unicodeobject.c [hunk 3]
```diff
@@ -5869,6 +5869,23 @@ PyUnicode_AsUTF16String(PyObject *unicode)
     return _PyUnicode_EncodeUTF16(unicode, NULL, 0);
 }
 
+_PyUnicode_Name_CAPI *
```
Comment (vstinner) [inline Objects/unicodeobject.c:5872]: Is there a reason why this function is not marked as "static"?

Comment (erlend-aasland) [reply Objects/unicodeobject.c:5872]: It's used in Python/codecs.c also.

Comment (vstinner) [reply Objects/unicodeobject.c:5872]: Oh, I didn't see that it's exposed in Include/internal/pycore_ucnhash.h. Ok, fine.

**Pull request merged**